File: python/1-50/37. Sudoku Solver.py
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if not board:
            return
        total = []
        rownums = [[] for _ in range(9)]
        colnums = [[] for _ in range(9)]
        gridnums = [[] for _ in range(9)]
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == ".":
                    total.append((i, j))
                else:
                    temp = board[i][j]
                    rownums[i].append(temp)
                    colnums[j].append(temp)
                    gridnums[i / 3 * 3 + j / 3].append(temp)

        def helper(level, total):
            if len(total) == level:
                return True
            i, j = total[level]
            allcan = map(str, range(1, 10))
            candidiates = list(set(allcan) - set(rownums[i])
                - set(colnums[j]) - set(gridnums[i / 3 * 3 + j/3]))
            if not candidiates:
                return False
            for num in candidiates:
                board[i][j] = str(num)
                rownums[i].append(str(num))
                colnums[j].append(str(num))
                gridnums[i / 3 * 3 + j / 3].append(str(num))
                flag = helper(level + 1, total)
                if not flag:
                    board[i][j] = "."
                    rownums[i].pop()
                    colnums[j].pop()
                    gridnums[i / 3 * 3 + j / 3].pop()
                else:
                    return True
        helper(0, total)

File: cpp/1-50/37. Sudoku Solver.cpp
class Solution {
public:
	void solveSudoku(vector<vector<char>>& board) {
		vector<vector<int>> positions;
		for (int i = 0; i < board.size(); i++)
		{
			for (int j = 0; j < board[0].size(); j++)
			{
				if (board[i][j] == '.')
				{
					std::vector<int> temp;
					temp.push_back(i);
					temp.push_back(j);
					positions.push_back(temp);
				}
			}
		}
		helper(board, positions, 0);
	}
	bool helper(vector<vector<char>>& board, vector<vector<int>> positions, int index)
	{
		if (index == positions.size())
		{
			return validate(board);
		}
		vector<int> temp = positions[index];
		for (int i = 0; i < 9; i++)
		{
			board[temp[0]][temp[1]] = '1' + i;
			if(!validate(board))
			{
			    board[temp[0]][temp[1]] = '.';
			}
			else{
			if(helper(board, positions, index+1))
			    return true;
			board[temp[0]][temp[1]] = '.';
			}
		}
		return false;
	}
    bool validate(vector<vector<char>>& board) {
        bool used1[9][9] = {false};
		bool used2[9][9] = {false};
		bool used3[9][9] = {false};
		for (int i = 0; i < board.size(); i++)
		{
			for (int j = 0; j < board[0].size(); j++)
			{
			    if(board[i][j] != '.')
			    {
				int num = board[i][j] - '1';
				int k = i / 3 * 3 + j / 3;
				if (used1[i][num] | used2[j][num] | used3[k][num])
					return false;
				else
					used1[i][num] = used2[j][num] = used3[k][num] = true;
			    }
			}
		}
		return true;
    }
};